parlai/core/beam.py

```diff
@@ -0,0 +1,184 @@
+import torch
+import math
+
+class Beam(object):
+    def __init__(self, beam_size, min_length=3, padding_token=0, bos_token=1, eos_token=2, min_n_best=3, cuda=False):
+        """
+        Generic beam class. It keeps information about beam_size hypothesis.
+        :param beam_size: number of hypothesis in the beam
+        :param min_length: minimum length of the predicted sequence
+        :param padding_token: Set to 0 as usual in ParlAI
+        :param bos_token: Set to 1 as usual in ParlAI
+        :param eos_token: Set to 2 as usual in ParlAI
+        :param min_n_best: Beam will not be done unless this amount of finished hypothesis (with EOS) is done
+        :param cuda: What device to use for computations
+        """
+        self.beam_size = beam_size
+        self.min_length = min_length
+        self.eos = eos_token
+        self.bos = bos_token
+        self.pad = padding_token
+        if cuda is True:
+            self.device = 'cuda'
+        else:
+            self.device = 'cpu'
+        self.scores = torch.Tensor(self.beam_size).float().zero_().to(self.device)  # recent score for each hypo in the beam
+        self.all_scores = [torch.Tensor([0.0]*beam_size).to(self.device)]  # self.scores values per each time step
+        self.bookkeep = []  # backtracking id to hypothesis at previous time step
+        self.outputs = [torch.Tensor(self.beam_size).long().fill_(padding_token).to(self.device)]  # output tokens at each time step
+        self.finished = []  # keeps tuples (score, time_step, hyp_id)
+        self.eos_top = False
+        self.eos_top_ts = None
+        self.n_best_counter = 0
+        self.min_n_best = min_n_best
+
+    def get_output_from_current_step(self):
+        return self.outputs[-1]
+
+    def get_backtrack_from_current_step(self):
+        return self.bookkeep[-1]
+
+    def advance(self, softmax_probs):
+        voc_size = softmax_probs.size(-1)
```

Comment:
can you add a comment on the dimensions of softmax probs?

Reply:
done